File: ruby/lib/resp_bench/engine/key_generator.rb
# frozen_string_literal: true

require "concurrent"

module RespBench
  module Engine
    # Key generator for benchmark operations.
    # Produces identical key sequences as the Java reference implementation.
    #
    # Supports two generation algorithms:
    # - sequential_int: Keys 0, 1, 2, ... N-1 (wraps around)
    # - uniform_rand: Random keys using Java-compatible LCG
    class KeyGenerator
      # @param config [RespBench::Config::KeyspaceConfig] keyspace configuration
      # @param seed_override [Integer, nil] optional seed override for thread-local instances
      def initialize(config, seed_override: nil)
        @config = config
        @key_prefix = config.effective_key_prefix
        @key_size_bytes = config.key_size_bytes
        @keys_count = config.keys_count
        @seed = seed_override || config.seed_value
        @sequential_counter = Concurrent::AtomicFixnum.new(0)
        @random = JavaRandom.new(@seed)
        @mutex = Mutex.new # For thread-safe random access
      end

      # Generate the next key as a byte string
      # @return [String] the generated key (binary string)
      def next_key
        key_index = if @config.sequential_int?
                      @sequential_counter.increment - 1
                    else
                      @mutex.synchronize { @random.next_int(@keys_count) }
                    end

        key_index = key_index % @keys_count
        format_key(key_index)
      end

      # Reset the generator to initial state
      def reset
        @sequential_counter.value = 0
        @mutex.synchronize do
          @random.set_seed(@seed)
        end
      end

      # Factory method
      # @param config [RespBench::Config::KeyspaceConfig] keyspace configuration
      # @return [KeyGenerator] new key generator instance
      def self.create(config)
        new(config)
      end

      # Factory method with custom seed override.
      # Used to create thread-local key generators with unique but reproducible seeds.
      # @param config [RespBench::Config::KeyspaceConfig] keyspace configuration
      # @param seed [Integer] seed value to use instead of config's seed
      # @return [KeyGenerator] new key generator instance with overridden seed
      def self.create_with_seed(config, seed)
        new(config, seed_override: seed)
      end

      private

      # Format a key index into a full key string
      # Matches Java's String.format("%0Nd", keyIndex) behavior
      #
      # @param key_index [Integer] the key index
      # @return [String] formatted key string
      def format_key(key_index)
        # Calculate padding width: key_size_bytes - prefix length, minimum 1
        padding_width = [@key_size_bytes - @key_prefix.length, 1].max
        format_string = "%0#{padding_width}d"
        "#{@key_prefix}#{format_string % key_index}"
      end
    end
  end
end

File: python/src/resp_bench/engine/key_generator.py
"""Key generator producing sequences identical to the other engines.

- ``sequential_int``: keys 0, 1, 2, ... N-1, wrapping around.
- ``uniform_rand``: Java-LCG random keys (see :mod:`.java_random`).

Key formatting matches Java's ``String.format("%0Nd", index)``: the numeric
part is zero-padded to ``max(key_size_bytes - len(prefix), 1)`` digits.

Cross-worker semantics follow the Java reference:
- ``sequential_int`` uses a counter SHARED across all workers in a phase, so
  the workers collectively emit 0, 1, 2, ... (this is what populates the whole
  keyspace during a warmup/populate phase). Pass a shared :class:`Counter`.
- ``uniform_rand`` uses a per-worker RNG seeded ``base_seed + worker_index``.
"""

from __future__ import annotations

from typing import Optional

from ..config.keyspace_config import KeyspaceConfig
from .java_random import JavaRandom


class Counter:
    """A monotonic 0-based counter. Shared across a phase's workers.

    Safe to share across asyncio tasks: ``next_value`` performs its
    read-increment with no ``await`` in between, so it is atomic on the single
    event-loop thread.
    """

    def __init__(self, start: int = 0) -> None:
        self._value = start

    def next_value(self) -> int:
        current = self._value
        self._value += 1
        return current

    def reset(self) -> None:
        self._value = 0


class KeyGenerator:
    def __init__(
        self,
        config: KeyspaceConfig,
        seed_override: Optional[int] = None,
        sequential_counter: Optional[Counter] = None,
    ) -> None:
        self._config = config
        self._key_prefix = config.effective_key_prefix()
        self._key_size_bytes = config.key_size_bytes
        self._keys_count = config.keys_count
        self._seed = seed_override if seed_override is not None else config.seed_value()
        self._sequential_counter = sequential_counter or Counter()
        self._random = JavaRandom(self._seed)

    @classmethod
    def create(cls, config: KeyspaceConfig) -> "KeyGenerator":
        return cls(config)

    @classmethod
    def create_with_seed(
        cls,
        config: KeyspaceConfig,
        seed: int,
        sequential_counter: Optional[Counter] = None,
    ) -> "KeyGenerator":
        """Per-worker generator with a unique seed and an optional shared counter."""
        return cls(config, seed_override=seed, sequential_counter=sequential_counter)

    def next_key(self) -> str:
        if self._config.is_sequential_int():
            key_index = self._sequential_counter.next_value()
        else:
            key_index = self._random.next_int(self._keys_count)

        key_index %= self._keys_count
        return self._format_key(key_index)

    def reset(self) -> None:
        self._sequential_counter.reset()
        self._random.set_seed(self._seed)

    def _format_key(self, key_index: int) -> str:
        padding_width = max(self._key_size_bytes - len(self._key_prefix), 1)
        return f"{self._key_prefix}{key_index:0{padding_width}d}"
